apis/cpi/v1alpha1/cpiconfig_types.go

```diff
@@ -0,0 +1,205 @@
+	SecretName string `json:"secretName"`
+
+	// The namespace of secret that stores CPI configuration
+	// +kubebuilder:validation:Optional
+	SecretNamespace string `json:"secretNamespace"`
+}
+
+type ExtraArgs struct {
+	// External arguments for cloud provider
+	// +kubebuilder:validation:Optional
+	TLSCipherSuites string `json:"tls-cipher-suites"`
+}
+
+type VSphereCPI struct {
+	// The vSphere mode. Either `vsphereCPI` or `vsphereParavirtualCPI`. Default value is `vsphereCPI`
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:=vsphereCPI
+	// +kubebuilder:validation:Enum=vsphereCPI;vsphereParavirtualCPI
+	Mode string `json:"mode"`
+
+	// The IP address or FQDN of the vSphere endpoint
+	// +kubebuilder:validation:Required
+	Server string `json:"server"`
+
+	// The datacenter in which VMs are created/located
+	// +kubebuilder:validation:Required
+	Datacenter string `json:"datacenter"`
+
+	// Username used to access a vSphere endpoint
+	// +kubebuilder:validation:Required
+	Username string `json:"username"`
+
+	// Password used to access a vSphere endpoint
+	// +kubebuilder:validation:Required
+	Password string `json:"password"`
+
+	// The flag that disables TLS peer verification
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:=false
+	InsecureFlag bool `json:"insecureFlag"`
+
+	// The IP family configuration
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Enum="ipv4";"ipv6";"ipv4,ipv6";"ipv6,ipv4"
+	// +kubebuilder:default:=ipv4
+	IPFamily string `json:"ipFamily"`
+
+	// Internal VM network name
+	// +kubebuilder:validation:Optional
+	VMInternalNetwork string `json:"vmInternalNetwork"`
+
+	// External VM network name
+	// +kubebuilder:validation:Optional
+	VMExternalNetwork string `json:"vmExternalNetwork"`
+
+	// Internal VM network CIDR to be excluded
+	// +kubebuilder:validation:Optional
+	VMExcludeInternalNetworkSubnetCidr string `json:"vmExcludeInternalNetworkSubnetCidr"`
+
+	// External VM network CIDR to be excluded
+	// +kubebuilder:validation:Optional
+	VMExcludeExternalNetworkSubnetCidr string `json:"vmExcludeExternalNetworkSubnetCidr"`
+
+	CloudProviderExtraArgs ExtraArgs `json:"cloudProviderExtraArgs"`
+
+	NsxT NsxT `json:"nsxt"`
+
+	// HTTP proxy setting
+	// +kubebuilder:validation:Optional
+	HTTPProxy string `json:"http_proxy"`
+
+	// HTTPS proxy setting
+	// +kubebuilder:validation:Optional
+	HTTPSProxy string `json:"https_proxy"`
+
+	// No-proxy setting
+	// +kubebuilder:validation:Optional
+	NoProxy string `json:"no_proxy"`
+
+	// Used in vsphereParavirtual mode, defines the Cluster API versions. Default: run.tanzu.vmware.com/v1alpha2.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:=run.tanzu.vmware.com/v1alpha2
+	ClusterAPIVersion string `json:"clusterAPIVersion"`
+
+	// Used in vsphereParavirtual mode, defines the Cluster kind. Default: Cluster.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:=Cluster
+	ClusterKind string `json:"clusterKind"`
```

Comment:
Question: Why we need a param for clusterKind? Wouldn't it always be `Cluster`?

Reply:
there might be other CR kind for example `TanzuKubernetesCluster`

Reply:
Could you describe the use case for this?

Reply:
This is used for ownerRef
Do we still need to support TKGs? They use `TanzuKubernetesCluster` to deploy Cluster. hm, But in that case, they still has Cluster object.
I guess it's just for better visibility when using `TKC`. But yeah, we can also directly refer to Cluster